guardian/typerighter
# Streaming API endpoint

Add a new `checkStream` endpoint that is identical to `check`, but gives a chunked response.

The reponse contains multiple JSON documents separated with a `RECORD_SEPARATOR` character per the [json-seq spec.](https://en.wikipedia.org/wiki/JSON_streaming#Record_separator-delimited_JSON_2)

There are two advantages to this approach:

- It allows Typerighter to reason about the whole document in one go, opening the door to matchers that care about the entire doc.
- It's less work for the client by a factor of about two. Here's the total browser processor time in Chrome, for the default long read in the prosemirror-client, averaged over 3 attempts on a macbook pro 2019 in ms:

|Before|After|
|-|-|
|1614|784|

## Why not server-sent events, or websockets?

We explored both of these:
- Server sent events would be perfect but for the fact that the current Javascript API [(spec)](https://html.spec.whatwg.org/multipage/server-sent-events.html) does not allow you to specify a verb in the request, or provide a request body. This limits us to query params when sending text, which across browsers is about [2000 characters](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers) – unacceptably small for entire documents.
- Websockets fit the use case nicely, but the bidirectional communication once a connection is established is a larger surface area than we need. Sending an initial payload including the document, and streaming back a response that was considered complete at its end, was a better semantic fit. In addition to this, the integration Play provides with streams for chunked responses means we need to write less code – the websocket approach required an Actor-based implementation to handle incoming messages.

## Dev notes

I've confirmed in Wireshark that the chunks in the TCP correspond to the records in the stream. However, there's no guarantee that those chunks will be what browsers provide via their APIs – see the [sibling PR](https://github.com/guardian/prosemirror-typerighter/pull/214) in prosemirror-typerighter for details.

## How to test

The automated tests should pass, and cover the core stream behaviour, including successfully issuing responses, failures happening within the `matcherPool` stream as a result of backpressure, and failures caused by the stream created by `checkStream` specifically.

Spin up the prosemirror-typerighter client on [this branch](https://github.com/guardian/prosemirror-typerighter/tree/jsh/default-to-ws), which adds an adapter that handles a chunked response, and run a check against the default document. You should be able to see that a single request goes over the wire, and that responses stream back into the document over time. This change is a no-op – everything should work as it did.
